From 5936ead583f2c7c93f7177ff44f0abcfda212f01 Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Wed, 6 Feb 2013 00:06:45 +0800
Subject: [PATCH 02/43] ENGR00243333 Add showFPS feature in Surfaceflinger

This feature is for debugging system performance only.

Signed-off-by: Xianzhong <b07117@freescale.com>

Conflicts:
	services/surfaceflinger/SurfaceFlinger.h

Conflicts:
	services/surfaceflinger/SurfaceFlinger.h
---
 services/surfaceflinger/SurfaceFlinger.cpp | 27 +++++++++++++++++++++++++++
 services/surfaceflinger/SurfaceFlinger.h   |  4 ++++
 2 files changed, 31 insertions(+)

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index fdc3650..8bce788 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -139,6 +139,7 @@ SurfaceFlinger::SurfaceFlinger()
         mDebugRegion(0),
         mDebugDDMS(0),
         mDebugDisableHWC(0),
+        mDebugFps(0),
         mDebugDisableTransformHint(0),
         mDebugInSwapBuffers(0),
         mLastSwapBufferTime(0),
@@ -177,8 +178,13 @@ SurfaceFlinger::SurfaceFlinger()
             mDebugDDMS = 0;
         }
     }
+
+    property_get("debug.sf.showfps", value, "0");
+    mDebugFps = atoi(value);
+
     ALOGI_IF(mDebugRegion, "showupdates enabled");
     ALOGI_IF(mDebugDDMS, "DDMS debugging enabled");
+    ALOGI_IF(mDebugFps,  "showfps enabled");
 }
 
 void SurfaceFlinger::onFirstRef()
@@ -1237,6 +1243,24 @@ void SurfaceFlinger::doComposition() {
     postFramebuffer();
 }
 
+void SurfaceFlinger::debugShowFPS() const
+{
+    static int mFrameCount;
+    static int mLastFrameCount = 0;
+    static nsecs_t mLastFpsTime = 0;
+    static float mFps = 0;
+    mFrameCount++;
+    nsecs_t now = systemTime();
+    nsecs_t diff = now - mLastFpsTime;
+    if (diff > ms2ns(250)) {
+        mFps =  ((mFrameCount - mLastFrameCount) * float(s2ns(1))) / diff;
+        mLastFpsTime = now;
+        mLastFrameCount = mFrameCount;
+    }
+    // XXX: mFPS has the value we want
+    ALOGI("fps - %.2f",mFps);
+}
+
 void SurfaceFlinger::postFramebuffer()
 {
     ATRACE_CALL();
@@ -1286,6 +1310,9 @@ void SurfaceFlinger::postFramebuffer()
     uint32_t flipCount = getDefaultDisplayDevice()->getPageFlipCount();
     if (flipCount % LOG_FRAME_STATS_PERIOD == 0) {
         logFrameStats();
+
+    if (mDebugFps) {
+        debugShowFPS();
     }
 }
 
diff --git a/services/surfaceflinger/SurfaceFlinger.h b/services/surfaceflinger/SurfaceFlinger.h
index b3baadd..ed29269 100644
--- a/services/surfaceflinger/SurfaceFlinger.h
+++ b/services/surfaceflinger/SurfaceFlinger.h
@@ -506,6 +506,10 @@ private:
     nsecs_t mFrameBuckets[NUM_BUCKETS];
     nsecs_t mTotalTime;
     nsecs_t mLastSwapTime;
+
+    //enabled by debug.sf.showfps for debug purpose
+    int  mDebugFps;
+    void debugShowFPS() const;
 };
 
 }; // namespace android
-- 
1.8.0

